### PR TITLE
CDPS-262 homepage services

### PR DESCRIPTION
**Pull request by jamesgrant-moj**

### CODE REVIEW

File: server/controllers/homepageController.ts [hunk 1]
```diff
@@ -1,34 +1,53 @@
...
   public displayHomepage(): RequestHandler {
-    return async (req: Request, res: Response, next: NextFunction) => {
-      const { userRoles, activeCaseLoadId } = res.locals.user
+    return async (req: Request & CustomRequest, res: Response, next: NextFunction) => {
+      const { activeCaseLoadId } = res.locals.user
 
       // Search Section
       const errors = req.flash('errors')
-      const userHasGlobal = userHasRoles([Role.GlobalSearch], userRoles)
+      const userHasRoles = (roles: string[]) => hasAnyRole(res.locals.user.userRoles, roles)
```
Comment (davidmiddletonmoj) [inline server/controllers/homepageController.ts:27]: Seems confusing to create your own function with the same name as the existing function from `utils.ts`
I think it's better to call your new function something else.

File: server/controllers/homepageController.ts [hunk 2]
```diff
@@ -52,4 +71,391 @@ export default class HomepageController {
...
+
+  getTasks(
+    activeCaseLoadId: string,
+    locations: Location[],
+    staffId: string,
+    whereaboutsConfig: WhereAboutsConfig,
+    keyworkerPrisonStatus: KeyWorkerPrisonStatus,
+    userRoles: string[],
+  ) {
+    const userHasRoles = (roles: string[]) => hasAnyRole(userRoles, roles)
```
Comment (davidmiddletonmoj) [inline server/controllers/homepageController.ts:113]: See previous comments.
Seems confusing to create your own function with the same name as the existing function from `utils.ts`
I think it's better to call your new function something else.

File: server/controllers/homepageController.ts [hunk 3]
```diff
@@ -52,4 +71,391 @@ export default class HomepageController {
...
+
+  getServices(req: CustomRequest, res: Response, next: NextFunction) {
+    const { whereaboutsMaintenanceMode } = config.app
+    const { keyworkerMaintenanceMode } = config.app
+
+    let whereaboutsConfig
+    if (whereaboutsMaintenanceMode) {
+      whereaboutsConfig = { enabled: false }
+    } else {
+      whereaboutsConfig = { enabled: true }
```
Comment (davidmiddletonmoj) [inline server/controllers/homepageController.ts:83]: DPS has calls to APIs to determine `whereaboutsConfig` and `keyworkerPrisonStatus` - do we not need these as well?

File: server/utils/nunjucksSetup.ts [hunk 4]
```diff
@@ -53,6 +54,8 @@ export default function nunjucksSetup(app: express.Express, applicationInfo: App
   njkEnv.addGlobal('userHasRoles', userHasRoles)
   njkEnv.addGlobal('userHasAllRoles', userHasAllRoles)
 
+  njkEnv.addGlobal('baseUrl', baseUrl)
```
Comment (davidmiddletonmoj) [inline server/utils/nunjucksSetup.ts:57]: Can probably get rid of this if we're not going to be using service sublinks

File: server/utils/utils.ts [hunk 5]
```diff
@@ -92,3 +93,8 @@ export const addDefaultSelectedValue = (items: SelectItem[], text: string, value
 
 export const asSelectItem = (items: Record<string, string>[], text: string, value?: string) =>
   items && items.map(entry => ({ value: entry[value || text], text: entry[text] }))
+
```
Comment (davidmiddletonmoj) [inline server/utils/utils.ts:96]: Can probably get rid of this if we're not going to be using service sublinks

**Comment by jamesgrant-moj** [comment]
Thanks for taking a look @davidmiddletonmoj, I have made all the changes apart from the last comment which is just white space.

There is a 401 error when calling 
https://keyworker-api-dev.prison.service.justice.gov.uk/swagger-ui/index.html#/key-worker/getPrisonMigrationStatus

The keyworker API as the following authorise roles. I am not sure if we need to add these to user or app level credentials.
@PreAuthorize("hasRole('SYSTEM_USER')")
@PreAuthorize("hasRole('KW_MIGRATION')")

I have also skipped two newly created unit tests for the APIs - I can not get them to pass.

File: server/controllers/homepageController.ts [hunk 6]
```diff
@@ -1,34 +1,53 @@
...
 /**
  * Parse requests for case notes routes and orchestrate response
  */
 export default class HomepageController {
   constructor(private readonly homepageService: HomepageService, private readonly todayCache: HmppsCache) {}
 
   public displayHomepage(): RequestHandler {
-    return async (req: Request, res: Response, next: NextFunction) => {
-      const { userRoles, activeCaseLoadId } = res.locals.user
+    return async (req: Request & CustomRequest, res: Response, next: NextFunction) => {
```
[the remark this answers is not in the document]
Comment (davidmiddletonmoj) [reply server/controllers/homepageController.ts:18]: Not sure if this was intentionally kept in?

File: server/views/partials/homepage/services.njk [hunk 7]
```diff
@@ -1 +1,15 @@
-<div class="tbd">Services</div>
\ No newline at end of file
+<section data-qa="homepage-services-section" data-qa="homepage-services-section">
+    <h2 data-qa="homepage-services-heading">Services</h3>
+    <div class="hmpps-services">
+        {% for service in services %}
+            <div class="hmpps-service-item">
+                <hr class="govuk-section-break govuk-section-break--m govuk-section-break--visible govuk-!-static-margin-top-0" />
+                <div class="govuk-grid-row">
+                    <div class="govuk-grid-column-three-quarters"><a href="{{ service.href }} " class="govuk-link govuk-!-font-weight-bold">{{ service.heading}}</a></div>
```
Comment (davidmiddletonmoj) [inline server/views/partials/homepage/services.njk:8]: Needs `govuk-link--no-visited-state`

**Comment by davidmiddletonmoj** [comment]
> There is a 401 error when calling https://keyworker-api-dev.prison.service.justice.gov.uk/swagger-ui/index.html#/key-worker/getPrisonMigrationStatus
> 
> The keyworker API as the following authorise roles. I am not sure if we need to add these to user or app level credentials. @PreAuthorize("hasRole('SYSTEM_USER')") @PreAuthorize("hasRole('KW_MIGRATION')")

Apparently we should be moving away from using `SYSTEM_USER`, so probably need to get the `KW_MIGRATION` role added to the app client.

**Comment by jamesgrant-moj** [comment]
Hi @davidmiddletonmoj, I have updated the PR.

Request & CustomRequest type was intentional in order to get it to work.

**Comment by jamesgrant-moj** [comment]
@davidmiddletonmoj ready for you to review again.

**Review by davidmiddletonmoj — APPROVED** [review]

**Pull request merged**